webapp/src/compiler.ts

```diff
@@ -386,15 +386,76 @@ export function getApisInfoAsync() {
 }
 
 export function getBlocksAsync(): Promise<pxtc.BlocksInfo> {
-    return cachedBlocks
-        ? Promise.resolve(cachedBlocks)
-        : getApisInfoAsync().then(info => {
+    if (!cachedBlocks) {
+        // Used packaged info
+        const bannedCategories = pkg.mainPkg.resolveBannedCategories();
+        const apis = getApiInfo(pkg.mainEditorPkg(), pxt.getBundledApiInfo())
+        if (apis) {
+            return Promise.resolve(cachedBlocks = pxtc.getBlocksInfo(apis, bannedCategories));
+        }
+
+        return getApisInfoAsync().then(info => {
             const bannedCategories = pkg.mainPkg.resolveBannedCategories();
             cachedBlocks = pxtc.getBlocksInfo(info, bannedCategories);
             return cachedBlocks;
         });
+    }
+
+    return Promise.resolve(cachedBlocks);
+}
+
+interface BundledPackage {
+    dirname: string;
+    config: pxt.PackageConfig;
+    files: pxt.Map<string>;
 }
 
+function getApiInfo(project: pkg.EditorPackage, bundled: pxt.Map<pxt.PackageApiInfo>) {
+    const bundledPackages: BundledPackage[] = pxt.appTarget.bundleddirs.map(dirname => {
+        const pack = pxt.appTarget.bundledpkgs[dirname.substr(dirname.lastIndexOf("/") + 1)];
+
+        if (!pack) return null;
+
+        return {
+            dirname,
+            files: pack,
+            config: JSON.parse(pack[pxt.CONFIG_NAME]) as pxt.PackageConfig
```

Comment:
pxt.json may be missing or invalid format

Reply:
it's not the only spot we make assumptions that's it's there but it would be good to know if the service will recover if pxt.json is invalid.

Reply:
I think this one is coming from pxtarget.json so we should be fine